File: apps/web/src/features/bee/voice-api.ts
import { z } from 'zod'

// BeeGreat keeps its Flue worker on a dedicated local port to avoid Vite collisions.
export const AGENT_URL =
  import.meta.env.VITE_AGENT_URL ?? 'http://localhost:3583'

type GetToken = () => Promise<string | null>

const voiceErrorBody = z.object({ error: z.string().optional() })
const transcriptionBody = z.object({ text: z.string() })
const speechBody = z.object({
  audio: z.string(),
  mimeType: z.string().optional(),
})
const realtimeTokenBody = z.object({
  token: z.string(),
  expiresAt: z.number(),
})

async function authHeaders(getToken: GetToken) {
  const token = await getToken()
  const headers: Record<string, string> = {}
  if (token) headers.authorization = `Bearer ${token}`
  return headers
}

async function readError(response: Response, fallback: string) {
  const body = voiceErrorBody.safeParse(
    await response.json().catch(() => null),
  )
  return (
    (body.success ? body.data.error : undefined) ??
    `${fallback} (HTTP ${response.status})`
  )
}

export async function transcribeBlob(blob: Blob, getToken: GetToken) {
  const response = await fetch(`${AGENT_URL}/voice/transcribe`, {
    method: 'POST',
    headers: {
      'content-type': blob.type || 'audio/webm',
      ...(await authHeaders(getToken)),
    },
    body: blob,
  })
  if (!response.ok) {
    throw new Error(await readError(response, 'Transcription failed.'))
  }
  const result = transcriptionBody.parse(await response.json())
  return result.text.trim()
}

export async function synthesizeSpeech(text: string, getToken: GetToken) {
  const response = await fetch(`${AGENT_URL}/voice/speak`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      ...(await authHeaders(getToken)),
    },
    body: JSON.stringify({ text }),
  })
  if (!response.ok) {
    throw new Error(await readError(response, 'Speech synthesis failed.'))
  }
  return speechBody.parse(await response.json())
}

export async function createRealtimeVoiceToken(getToken: GetToken) {
  const response = await fetch(`${AGENT_URL}/voice/realtime-token`, {
    method: 'POST',
    headers: await authHeaders(getToken),
  })
  if (!response.ok) {
    throw new Error(
      await readError(response, 'Conversational voice could not start.'),
    )
  }
  return realtimeTokenBody.parse(await response.json())
}

File: apps/android/app/src/main/kotlin/com/beegreat/app/voice/VoiceApi.kt
package com.beegreat.app.voice

import com.beegreat.app.BuildConfig
import com.beegreat.flue.FlueJson
import com.beegreat.flue.HeaderProvider
import com.beegreat.flue.await
import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

@Serializable private data class TranscriptionResponse(val text: String)

@Serializable private data class SpeechResponse(val audio: String)

@Serializable data class RealtimeToken(val token: String, val expiresAt: Double, val websocketUrl: String)

/** The worker voice routes (transcribe, speak, realtime token). Port of `apps/mobile/src/lib/voice-api.ts`. */
class VoiceApi(private val http: OkHttpClient, private val headers: HeaderProvider, private val agentUrl: String = BuildConfig.AGENT_URL) {
  /** Raw bytes with the audio content type; the worker builds the multipart request itself. */
  suspend fun transcribe(file: File, mimeType: String = "audio/m4a"): String {
    val response = send(Request.Builder().url("$agentUrl/voice/transcribe").post(file.readBytes().toRequestBody(mimeType.toMediaType())))
    return response.use {
      if (!it.isSuccessful) throw VoiceApiException(readError(it, "Transcription failed."))
      FlueJson.decodeFromString(TranscriptionResponse.serializer(), it.body.string()).text.trim()
    }
  }

  /** MP3 bytes for [text]; the caller owns the cache file. */
  suspend fun speak(text: String): ByteArray {
    val body = FlueJson.encodeToString(kotlinx.serialization.json.JsonObject.serializer(), kotlinx.serialization.json.buildJsonObject { put("text", kotlinx.serialization.json.JsonPrimitive(text)) })
    val response = send(Request.Builder().url("$agentUrl/voice/speak").post(body.toRequestBody("application/json".toMediaType())))
    return response.use {
      if (!it.isSuccessful) throw VoiceApiException(readError(it, "Speech synthesis failed."))
      android.util.Base64.decode(FlueJson.decodeFromString(SpeechResponse.serializer(), it.body.string()).audio, android.util.Base64.DEFAULT)
    }
  }

  suspend fun realtimeToken(): RealtimeToken {
    val response = send(Request.Builder().url("$agentUrl/voice/realtime-token").post(ByteArray(0).toRequestBody(null)))
    return response.use {
      if (!it.isSuccessful) throw VoiceApiException(readError(it, "Conversational voice could not start."))
      FlueJson.decodeFromString(RealtimeToken.serializer(), it.body.string())
    }
  }

  private suspend fun send(builder: Request.Builder): Response {
    for ((key, value) in headers()) builder.header(key, value)
    return http.newCall(builder.build()).await()
  }

  /** Prefers the worker's `{ error }` message so the UI shows the real cause. */
  private fun readError(response: Response, fallback: String): String {
    val message = runCatching { FlueJson.parseToJsonElement(response.body.string()).jsonObject["error"]?.jsonPrimitive?.content }.getOrNull()
    return message ?: "$fallback (HTTP ${response.code})"
  }
}

class VoiceApiException(message: String) : RuntimeException(message)
